protocol/v2/ssv/validator/committee_guard.go

```diff
@@ -0,0 +1,70 @@
+package validator
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	spectypes "github.com/ssvlabs/ssv-spec/types"
+)
+
+// CommitteeDutyGuard helps guarantee exclusive execution of one duty per validator
+// and non-execution of stopped validators.
+type CommitteeDutyGuard struct {
+	duties map[spectypes.BeaconRole]map[spectypes.ValidatorPK]phase0.Slot
+	mu     sync.RWMutex
+}
+
+func NewCommitteeDutyGuard() *CommitteeDutyGuard {
+	return &CommitteeDutyGuard{
+		duties: map[spectypes.BeaconRole]map[spectypes.ValidatorPK]phase0.Slot{
+			spectypes.BNRoleAttester:      {},
+			spectypes.BNRoleSyncCommittee: {},
+		},
+	}
+}
+
+// StartDuty records a started duty. If a duty is already running at the same or higher slot, it returns an error.
+func (a *CommitteeDutyGuard) StartDuty(role spectypes.BeaconRole, validator spectypes.ValidatorPK, slot phase0.Slot) error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	duties, ok := a.duties[role]
+	if !ok {
+		return fmt.Errorf("unsupported role %d", role)
+	}
+	runningSlot, exists := duties[validator]
+	if exists && runningSlot >= slot {
+		return fmt.Errorf("duty already running at slot %d", runningSlot)
+	}
```

Comment:
if `exists && slot > running slot` shouldn't we stop the existing duty before starting this one?

Reply:
it does, indirectly

in case the given slot is higher than running slot (`exists && slot > running slot`), then it will replace it in the map, after which `ValidDuty` would return an error upon checking the previously running slot

maybe we should rename the methods or update comments to make it clear

Reply:
I get it now. And it seems that previous approach was doing it pretty similar to this...

Reply:
maybe you guys want to also mention `slot` in the error?
Or maybe your logger does this already?